shell/platform/android/io/flutter/plugin/editing/FlutterTextUtils.java

```diff
@@ -0,0 +1,184 @@
+
+  public boolean isVariationSelector(int codePoint) {
+    return flutterJNI.nativeFlutterTextUtilsIsVariationSelector(codePoint);
+  }
+
+  public boolean isRegionalIndicatorSymbol(int codePoint) {
+    return flutterJNI.nativeFlutterTextUtilsIsRegionalIndicator(codePoint);
+  }
+
+  public boolean isTagSpecChar(int codePoint) {
+    return 0xE0020 <= codePoint && codePoint <= 0xE007E;
+  }
+
+  public boolean isKeycapBase(int codePoint) {
+    return ('0' <= codePoint && codePoint <= '9') || codePoint == '#' || codePoint == '*';
+  }
+
+  /**
+   * Start offset for backspace key or moving left from the current offset. Same methods are also
+   * included in Android APIs but they don't work as expected in API Levels lower than 24. Reference
+   * for the logic in this code is the Android source code.
+   *
+   * @see <a target="_new"
+   *     href="https://android.googlesource.com/platform/frameworks/base/+/refs/heads/android10-s3-release/core/java/android/text/method/BaseKeyListener.java#111">https://android.googlesource.com/platform/frameworks/base/+/refs/heads/android10-s3-release/core/java/android/text/method/BaseKeyListener.java#111</a>
+   */
+  public int getOffsetBefore(CharSequence text, int offset) {
+    if (offset <= 1) {
+      return 0;
+    }
+
+    int codePoint = Character.codePointBefore(text, offset);
+    int deleteCharCount = Character.charCount(codePoint);
+    int lastOffset = offset - deleteCharCount;
+
+    if (lastOffset == 0) {
+      return 0;
+    }
+
+    // Line Feed
+    if (codePoint == LINE_FEED) {
+      codePoint = Character.codePointBefore(text, lastOffset);
+      if (codePoint == CARRIAGE_RETURN) {
+        ++deleteCharCount;
+      }
+      return offset - deleteCharCount;
+    }
+
+    // Flags
+    if (isRegionalIndicatorSymbol(codePoint)) {
+      codePoint = Character.codePointBefore(text, lastOffset);
+      lastOffset -= Character.charCount(codePoint);
+      int regionalIndicatorSymbolCount = 1;
+      while (lastOffset > 0 && isRegionalIndicatorSymbol(codePoint)) {
+        codePoint = Character.codePointBefore(text, lastOffset);
+        lastOffset -= Character.charCount(codePoint);
+        regionalIndicatorSymbolCount++;
+      }
+      if (regionalIndicatorSymbolCount % 2 == 0) {
+        deleteCharCount += 2;
+      }
+      return offset - deleteCharCount;
+    }
+
+    // Keycaps
+    if (codePoint == COMBINING_ENCLOSING_KEYCAP) {
+      codePoint = Character.codePointBefore(text, lastOffset);
+      lastOffset -= Character.charCount(codePoint);
+      if (lastOffset > 0 && isVariationSelector(codePoint)) {
+        int tmpCodePoint = Character.codePointBefore(text, lastOffset);
+        if (isKeycapBase(tmpCodePoint)) {
+          deleteCharCount += Character.charCount(codePoint) + Character.charCount(tmpCodePoint);
+        }
+      } else if (isKeycapBase(codePoint)) {
+        deleteCharCount += Character.charCount(codePoint);
+      }
+      return offset - deleteCharCount;
+    }
+
```

Comment:
Since you have unrolled the state machine, can you add some notation here indicating that the following few if statements are meant to act like a fall through switch and do not return like the if statements above?

The separation will make the code much more parse-able.

Reply:
Done